### PR TITLE
Constraints must use `Self`

### PR DESCRIPTION
**Pull request by josh11b**
Require `impl as` constraints in an `interface` or `constraint` definition to mention `Self` implicitly or explicitly. Require `where` clauses to refer to `.Self` directly, or through a designator like `.Foo`.

### CODE REVIEW

File: docs/design/generics/details.md [hunk 1]
```diff
@@ -3713,6 +3760,10 @@ class Array(T:! Type, template N:! Int) {
 }
 ```
 
+The `impl` declarations in the body of a `class` definition must be for the
+class being defined. It is an error to define `impl i32 as Printable` inside
```
Comment (zygoloid) [inline docs/design/generics/details.md]:
```suggestion
The `impl` declarations in the body of a `class` definition must be for the
class being defined. It is an error to declare `impl i32 as Printable` inside
```

File: docs/design/generics/details.md [hunk 2]
```diff
@@ -3713,6 +3760,10 @@ class Array(T:! Type, template N:! Int) {
 }
 ```
 
+The `impl` declarations in the body of a `class` definition must be for the
```
Comment (zygoloid) [inline docs/design/generics/details.md]: Does this apply only to internal `impl`s, or also to `external impl` declarations?

File: proposals/p2376.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+-   `Type where Vector(.Self) is Sortable`
+-   `Addable where i32 is AddableWith(.Result)`
+
+`impl as` declarations in interfaces and named constraints must always involve
+`Self`:
+
+-   Can be the implicit `Self` when no type is specified, as in `impl as ...`,
+    or the equivalent declarations with `Self` declared explicitly, as in
+    `impl Self as ...`
+-   Can be an argument a type. The type can be what is to the left of the `as`,
```
Comment (zygoloid) [inline proposals/p2376.md]: First sentence seems to be missing a word.

**Review by zygoloid — COMMENTED** [review]
Looks good to me; I think this accurately captures the discussions and conclusions.

File: docs/design/generics/details.md (hunk 2 above)
Comment (josh11b) [reply docs/design/generics/details.md]: I would say both, but I'm wondering if you have a use case for `external impl` of some other type? The only case I can think of it being useful is if you have an interface that requires `impl T as ComparableWith(Self)` you might want to do that external impl there.

File: proposals/p2376.md (hunk 3 above)
Comment (josh11b) [reply proposals/p2376.md]: added "to"

File: docs/design/generics/details.md [hunk 4]
```diff
@@ -2879,6 +2885,47 @@ fn Double[T:! Mul where i32 is As(.Self)](x: T) -> T {
...
+
+```
+// ❌ Error: `where T is B` does not use `.Self` or a designator
+external impl forall [T:! Type] T as A where T is B {}
+// ✅ Allowed
+external impl forall [T:! Type where .Self is B] T as A {}
+// ✅ Allowed
+external impl forall [T:! B] T as A {}
+```
+
```
Comment (chandlerc) [inline docs/design/generics/details.md:2924]: nit: Maybe add a super short summary (one sentence even) of the motivation here, based on the more detailed form you have in the proposal?

**Review by chandlerc — COMMENTED** [review]
(drive-by comment, don't block on me -- all of this looks nice)

File: docs/design/generics/details.md (hunk 2 above)
Comment (josh11b) [reply docs/design/generics/details.md]: In person we discussed this and came to the conclusion that there are use cases for `external impl`s that should have access to private members, such as conversion from <some other type> to `Self`, or implementing <some other type> `+ Self`.

File: docs/design/generics/details.md (hunk 4 above)
Comment (josh11b) [reply docs/design/generics/details.md:2924]: I added a sentence. Does that cover everything you wanted?

**Review by zygoloid — APPROVED** [review]
Leads are happy.

**Pull request merged**